### PR TITLE
WIP: Update deque

### PR DESCRIPTION
**Pull request by mattytrentini**
MicroPython's built-in deque lacks _many_ features but is implemented in C so is very lean and fast. 
The micropython-lib version is implemented in Python and has more features - but should have even more. This is a start toward achieving feature-parity with CPython's deque.

### CODE REVIEW

**Comment by mattytrentini** [comment]
This could possibly be merged as-is (I had to extend deque to supporting porting the [progress library](https://github.com/mattytrentini/progress/tree/micropython-support) to MicroPython) but I'm going to try and add [all the other features of deque](https://docs.python.org/3/library/collections.html#collections.deque) as well as use as much as possible of the [CPython test cases](https://github.com/python/cpython/blob/main/Lib/test/test_deque.py).

File: python-stdlib/collections.deque/collections/deque.py [hunk 1]
```diff
@@ -1,28 +1,44 @@
...
-    def popright(self):
-        return self.q.pop()
-
     def pop(self):
         return self.q.pop()
 
     def append(self, a):
         self.q.append(a)
+        if self.__maxlen is not None and len(self.q) > self.__maxlen:
+            self.popleft()
```
Comment (dpgeorge) [inline python-stdlib/collections.deque/collections/deque.py:23]: you could do this bit before appending, to use less memory for the list

**Comment by BrianPugh** [comment]
chiming in that this helped make some of my cpython code be micropython-compatible.

**Comment by mattytrentini** [comment]
Note that the built-in deque has had some substantial improvements, see:

https://github.com/micropython/micropython/commit/7dff38fdc190d7b731fad8319d2ae8aa13fde18a

There is little reason to use this Python version now - which is great!

**Comment by dpgeorge** [comment]
Now that the core `collections.deque` class is more fully implemented, I think we should just remove the `deque` implementation here.  That will prevent any confusion, everyone will just use the built-in version.

**Comment by mattytrentini** [comment]
I agree; the differences are minimal, so I'll close this. Fantastic to see the built-in deque improved!

**Pull request closed without merging**